<?xml version="1.0"?>
<robot name="my_robot" xmlns:xacro="http://www.ros.org/wiki/xacro">
  
  
  
  <!-- Define common material -->
  <xacro:property name="gray_color" value="0.7 0.7 0.7 1"/>
  <!-- Macro for a mesh link -->
  <xacro:macro name="box_link" params="link_name mass ix iy iz xyz rpy">
    <link name="${link_name}">
      <inertial>
        <origin xyz="${xyz}" rpy="${rpy}"/>
        <mass value="${mass}"/>
        <inertia ixx="${ix}" ixy="0" ixz="0" iyy="${iy}" iyz="0" izz="${iz}"/>
      </inertial>
      <visual>
        <origin xyz="${xyz}" rpy="${rpy}"/>
        <geometry>
          <box size="0.5 0.95 0.5"/>
        </geometry>
        <material name="gray">
          <color rgba="${gray_color}"/>
        </material>
      </visual>
      <collision>
        <origin xyz="${xyz}" rpy="${rpy}"/>
        <geometry>
          <box size="0.95 0.5 0.5"/>
        </geometry>
      </collision>
    </link>
  </xacro:macro>



  <!-- Macro for a cylinder link (wheel) -->
  <xacro:macro name="cylinder_link" params="link_name mass radius length ix iy iz xyz rpy">
    <link name="${link_name}">
      <inertial>
        <origin xyz="${xyz}" rpy="${rpy}"/>
        <mass value="${mass}"/>
        <inertia ixx="${ix}" ixy="0" ixz="0" iyy="${iy}" iyz="0" izz="${iz}"/>
      </inertial>
      <visual>
        <origin xyz="${xyz}" rpy="${rpy}"/>
        <geometry>
          <cylinder radius="${radius}" length="${length}"/>
        </geometry>
        <material name="gray">
          <color rgba="0.5 0.5 0.5 1"/>
        </material>
      </visual>
      <collision>
        <origin xyz="${xyz}" rpy="${rpy}"/>
        <geometry>
          <cylinder radius="${radius}" length="${length}"/>
        </geometry>
      </collision>
    </link>
  </xacro:macro>



  <!-- Link 0 (Main Body) -->
  <xacro:box_link 
    link_name="base_link" 
    mass="58.7" 
    ix="0.166667" 
    iy="0.166667" 
    iz="0.166667" 
    xyz="0 0 .3" 
    rpy="0 0 0"/>
  <!-- Front Left Wheel -->
  <xacro:cylinder_link 
    link_name="wheel_link_front_left" 
    mass="1.0" 
    radius="0.125" 
    length="0.11" 
    ix="0.001" 
    iy="0.001" 
    iz="0.001" 
    xyz="0 0 0" 
    rpy="1.5708 0 0"/>
  <xacro:cylinder_link 
    link_name="wheel_link_front_right" 
    mass="1.0" 
    radius="0.125" 
    length="0.11" 
    ix="0.001" 
    iy="0.001" 
    iz="0.001" 
    xyz="0 0 0" 
    rpy="1.5708 0 0"/>
  <xacro:cylinder_link 
    link_name="wheel_link_back_left" 
    mass="1.0" 
    radius="0.125" 
    length="0.11" 
    ix="0.001" 
    iy="0.001" 
    iz="0.001" 
    xyz="0 0 0" 
    rpy="1.5708 0 0"/>
  <xacro:cylinder_link 
    link_name="wheel_link_back_right" 
    mass="1.0" 
    radius="0.125" 
    length="0.110" 
    ix="0.001" 
    iy="0.001" 
    iz="0.001" 
    xyz="0 0 0" 
    rpy="1.5708 0 0"/>



  <!-- Joint connecting front left wheel to the main body -->
  <joint name="wheel_joint_front_left" type="continuous">
    <parent link="base_link"/>
    <child link="wheel_link_front_left"/>
    <origin xyz="0.305 .4 0" rpy="0 0 1.5708"/>
    <axis xyz="0 1 0"/>
    <limit effort="10" velocity="5.0"/>
  </joint>
  <joint name="wheel_joint_front_right" type="continuous">
    <parent link="base_link"/>
    <child link="wheel_link_front_right"/>
    <origin xyz="-0.305 .4 0" rpy="0 0 1.5708"/>
    <axis xyz="0 1 0"/>
    <limit effort="10" velocity="5.0"/>
  </joint>
  <joint name="wheel_joint_back_left" type="continuous">
    <parent link="base_link"/>
    <child link="wheel_link_back_left"/>
    <origin xyz="0.305 -.4 0" rpy="0 0 1.5708"/>
    <axis xyz="0 1 0"/>
    <limit effort="10" velocity="5.0"/>
  </joint>
  <joint name="wheel_joint_back_right" type="continuous">
    <parent link="base_link"/>
    <child link="wheel_link_back_right"/>
    <origin xyz="-0.305 -.4 0" rpy="0 0 1.5708"/>
    <axis xyz="0 1 0"/>
    <limit effort="10" velocity="5.0"/>
  </joint>



  <!-- Transmission for wheel_joint_back_right -->
  <transmission name="front_left_wheel_transmission">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="wheel_joint_front_left">
      <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
    </joint>
  </transmission>
  <transmission name="front_right_wheel_transmission">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="wheel_joint_front_right">
      <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
    </joint>
  </transmission>
  <transmission name="back_left_wheel_transmission">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="wheel_joint_back_left">
      <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
    </joint>
  </transmission>
  <transmission name="back_right_wheel_transmission">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="wheel_joint_back_right">
      <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
    </joint>
  </transmission>
</robot>